File: py/p7.py
#!/usr/bin/env python
#
# Finds the 10,001st prime number.
#
# (c) 2013 Matthew Rheaume

import math

def is_prime(n):
	for i in xrange(2, int(math.sqrt(n)) + 1):
		if (n % i == 0):
			return False

	return True

def solve():
	count = 0
	n = 1

	while (count < 10001):
		n += 1

		if (is_prime(n)):
			count += 1

	return n

if __name__ == "__main__":
	print solve()

File: hs/p7.hs
-- Finds the 10,001st prime number.
--
-- (c) 2013 Matthew Rheaume

module Main where

intSqrt :: Int -> Int
intSqrt = ceiling . sqrt . fromIntegral

isPrime :: Int -> Bool
isPrime 2 = True
isPrime n = foldr checkDivisor True [2..(intSqrt n)] where
	checkDivisor k = (n `mod` k /= 0 &&)

findPrime :: Int -> Int
findPrime count = checkPrimes count 2 where
	checkPrimes 0 n = n - 1
	checkPrimes k n =
		if (isPrime n) then
			checkPrimes (k - 1) (n + 1)
		else
			checkPrimes k (n + 1)

solve :: Int
solve = findPrime 10001

main :: IO ()
main = putStrLn $ show $ solve
